### PR TITLE
ADD: large batch support

### PR DESCRIPTION
**Pull request by madisonmay**
Reviewers: 
----------

- [ ] @sihrc 

What's the purpose of the PR?
-----------------------------

Adds the option to pass in a `batch_size` argument to our API that automatically groups large requests into smaller ones when sending data off to our API for analysis.

Description of changes
----------------------

Addition of `large_batch_support` decorator.
Addition of unittests for decorator.
Addition of integration tests for the feature.

Notes to reviewers
------------------

There are still a few unknowns that need to be answered before this ships.   Mainly, I'm curious how we should deal with error handling in the case that one of the batches fails.  UX thoughts @sihrc?

Keep in mind while reviewing code:
----------------------------------
- Is relevant code tested?
- Are added functions/methods documented?
- Separation of concerns (SOC)
- Don't repeat yourself (DRY) 
- Limit the number of positional arguments
- Is function/method length reasonable?
- Can code be broken down into smaller components?
- Where do added functions/methods belong?
- Are variable names descriptive?
- Are errors handled appropriately?
- Can logic be simplified?
- Does the code make sense in context? (expand the diff)

### CODE REVIEW

File: indicoio/text/sentiment.py [hunk 1]
```diff
@@ -48,4 +49,4 @@ def sentiment_hq(text, cloud=None, batch=False, api_key=None, version=None, **kw
     :rtype: Float
     """
     url_params = {"batch": batch, "api_key": api_key, "version": version}
-    return api_handler(text, cloud=cloud, api="sentimenthq", url_params=url_params, **kwargs)
+    return api_handler(text, cloud=cloud, api="sentimenthq", batch_size=batch_size, url_params=url_params, **kwargs)
```
Comment (sihrc) [inline indicoio/text/sentiment.py]: we could probably just do batch_size=kwargs.get("batch_size") - although I guess including it in the signature has more .. transparency

File: indicoio/utils/api.py [hunk 2]
```diff
@@ -37,10 +63,33 @@ def api_handler(arg, cloud, api, url_params=None, **kwargs):
...
+    Optionally split up a single request into a series of requests
+    to ensure timely HTTP responses.
+
+    Could eventually speed up the time required to receive a response by
+    sending batches to the indico API concurrently
+    """
+    if batch_size:
+        results = []
+        for batch in batched(input_data, size=batch_size):
+            results.extend(send_request(batch, url, headers, kwargs))
```
Comment (sihrc) [inline indicoio/utils/api.py]: How are we thinking about handling errors? Let them be raised and exit out of batched or localize errors to each "batch" and include the IndicoError in the result list..

Comment (sihrc) [inline indicoio/utils/api.py]: The way it is now probably works, just it would cancel large batches.

**Comment by madisonmay** [comment]
Checklist:
- [ ] Handle case where batch_size is passed in but a single example is passed.

File: tests/test_batching.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        self.assertTrue(all([isinstance(el, float) for el in response]))
+
+    def test_batched_error_handling(self):
+        test_data = ["Terribly interesting test data."] * 100
+        test_data[98] = ""
+        with self.assertRaises(IndicoError):
+            response = sentiment_hq(test_data, batch_size=20)
+
+        # cleanup after ourselves
+        for file in glob.glob('indico-sentimenthq-*.json'):
```
Comment (sihrc) [inline tests/test_batching.py]: check contents of file?

**Pull request closed without merging**